### PR TITLE
Add offload support to Bark

### PR DESCRIPTION
**Pull request by ylacombe**
# What does this PR do?

Bark is a TTS model recently added by #24086. 

It is made of is of 4 main sub-models, which are sequentially called during generation (`BarkModel.generate`). When a sub-model is used, the other models remain unused, taking up a lot of space on the GPU.

This PR thus propose a simple, yet effective, handmade offloading of sub-models.

```python
from transformers import BarkModel, BarkProcessor

# no need to load the model onto GPU, it will be done inside the generate function
model = BarkModel.from_pretrained("suno/bark")
processor = BarkProcessor.from_pretrained("suno/bark")

# works if a GPU is available. Throws a warning if not, and uses default behavior.
device = "cuda"

# input must be put onto the right device, i.e onto the GPU.
input = processor("Hey, it's a test").to(device)

# one simple additional argument
output = model.generate(**input, offload_to_cpu = True)

# output is loaded onto GPU as well
```

With this PR, GPU footprint is around 60% lower, while being less than 10% slower, based on a benchmark I've done and that will be shared soon (`batch_size = 1` on a single TITAN RTX).

TODO:
- [x] write tests

## Who can review?

Hey @sanchit-gandhi and @amyeroberts, I'm tagging you because you were the ones reviewing  #24086 and because the changes are really small! 
I'm wondering whether the code I've written conforms to the transformer paradigm and whether I need to raise additional warnings or errors in extreme cases!
Many thanks!

### CODE REVIEW

**Comment by ylacombe** [comment]
Hi @sgugger, thank you for taking part! 
Diffusers pipeline is using [`accelerate's cpu_offload`](https://github.com/huggingface/accelerate/blob/69e4c3c54da3201eda288b500d138761e7a5221c/src/accelerate/big_modeling.py#L146) which itself is using [`accelerate's dispatch model`](https://github.com/huggingface/accelerate/blob/69e4c3c54da3201eda288b500d138761e7a5221c/src/accelerate/big_modeling.py#L146).  

The only issue here is that accelerate's cpu offload is done whenever calling the forward pass of the model. Here, some sub-model are auto-regressive so it would (if I'm not wrong) offload/onload over and over again while forwarding during the `generate` call. This would be sub-optimal and time-consuming, and would remove the benefits of my version of `offload_to_cpu`.

BTW, I'm open to suggestions on how to best make it work with accelerate, if it's possible! @muellerzr (hi !) or @sgugger , do you have some ideas?

**Comment by ylacombe** [comment]
Nice, it's indeed what I'm doing @sgugger , many thanks for your help! I'll adapt the PR

**Comment by patrickvonplaten** [comment]
Here an example: https://github.com/huggingface/diffusers/blob/fa356bd4da2593c2b91f76c1f63b6238249ec001/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L225

File: src/transformers/models/bark/modeling_bark.py [hunk 1]
```diff
@@ -1376,6 +1401,63 @@ def __init__(self, config):
...
+            raise ImportError("`enable_model_cpu_offload` requires `accelerate`.")
+
+        device = torch.device(f"cuda:{gpu_id}")
+
+        if self.device.type != "cpu":
+            self.to("cpu")
+            torch.cuda.empty_cache()  # otherwise we don't see the memory savings (but they probably exist)
+
+        # this layer is used outside the first foward pass of semantic so need to be loaded before semantic
+        self.semantic.input_embeds_layer, _ = cpu_offload_with_hook(self.semantic.input_embeds_layer, device)
```
Comment (ylacombe) [inline src/transformers/models/bark/modeling_bark.py:1444]: Here is a first tweak.

File: src/transformers/models/bark/modeling_bark.py [hunk 2]
```diff
@@ -1490,9 +1572,20 @@ def generate(
             **kwargs_fine,
         )
 
+        if hasattr(self, "fine_acoustics_hook") and self.fine_acoustics_hook is not None:
+            # Manually offload fine_acoustics to CPU
+            # and load codec_model to GPU
+            # since bark doesn't use codec_model forward pass
+            self.fine_acoustics_hook.offload()
+            self.codec_model = self.codec_model.to(self.device)
+
```
Comment (ylacombe) [inline src/transformers/models/bark/modeling_bark.py:1581]: And here is the second one.

**Review by ylacombe — COMMENTED** [review]
Hi @sgugger and @patrickvonplaten,
thanks for your help and for the different examples on how to use [`cpu_offload_with_hook`](https://github.com/huggingface/accelerate/blob/69e4c3c54da3201eda288b500d138761e7a5221c/src/accelerate/big_modeling.py#L194)!

At the end of the day and as you suggested and after having discussed with @SunMarc , I refactored my code using `cpu_offload_with_hook`, with two specific tweaks because some sub-modules are called without calling the forward method of the corresponding sub-models. I flagged the two tweaks in the following comments.

Note that the results of the benchmark I've shared in the PR introduction are still the same with this implementation. 

I've also added a `test_generate_end_to_end_with_offload` to make sure it's working as expected.

Many thanks for your help.

File: src/transformers/models/bark/modeling_bark.py [hunk 3]
```diff
@@ -1490,9 +1572,20 @@ def generate(
             **kwargs_fine,
         )
 
+        if hasattr(self, "fine_acoustics_hook") and self.fine_acoustics_hook is not None:
```
Comment (sgugger) [inline src/transformers/models/bark/modeling_bark.py:1575]:
```suggestion
        if getattr(self, "fine_acoustics_hook", None) is not None:
```
nit

File: src/transformers/models/bark/modeling_bark.py [hunk 4]
```diff
@@ -1490,9 +1572,20 @@ def generate(
...
+            # Manually offload fine_acoustics to CPU
+            # and load codec_model to GPU
+            # since bark doesn't use codec_model forward pass
+            self.fine_acoustics_hook.offload()
+            self.codec_model = self.codec_model.to(self.device)
+
         # 4. Decode the output and generate audio array
         audio = self.codec_decode(output)
 
+        if hasattr(self, "codec_model_hook") and self.codec_model_hook is not None:
```
Comment (sgugger) [inline src/transformers/models/bark/modeling_bark.py:1585]: Same here:
```suggestion
        if getattr(self, "codec_model_hook", None) is not None:
```

**Review by sgugger — APPROVED** [review]
Looks great to me!

**Comment by ylacombe** [comment]
@sgugger , thanks for the quick review! I've applied your last nit comments.

**Review by sanchit-gandhi — APPROVED** [review]
LGTM - thanks for working on this @ylacombe!

**Comment by ylacombe** [comment]
Hi @sgugger and @sanchit-gandhi , again thanks for the review! 
I've applied your last nit comments.
Don't hesitate to reach out to me if I need to refactor or improve something!

File: tests/models/bark/test_modeling_bark.py [hunk 5]
```diff
@@ -51,6 +52,8 @@
         BarkSemanticModel,
     )
 
+set_seed(42)
```
Comment (sanchit-gandhi) [inline tests/models/bark/test_modeling_bark.py:55]: Do we require setting the seed now for the tests? Have we lost determinism somewhere?

Comment (ylacombe) [reply tests/models/bark/test_modeling_bark.py]: Nice catch, it was unrequired!

Comment (sanchit-gandhi) [reply tests/models/bark/test_modeling_bark.py]: Cool - will merge when the CI is green

**Pull request merged**